packages/cardpay-sdk/sdk/scheduled-payment-module.ts

```diff
@@ -680,16 +682,189 @@ export default class ScheduledPaymentModule {
+    payAt: number | null,
+    recurringDayOfMonth?: number | null,
+    recurringUntil?: number | null,
+    onScheduledPaymentCreate?: (scheduledPaymentId: string) => unknown
+  ): Promise<void>;
+  async schedulePayment(
+    safeAddressOrScheduledPaymentId: string,
+    moduleAddress?: string,
+    tokenAddress?: string,
+    amount?: string,
+    payeeAddress?: string,
+    feeFixedUSD?: number,
+    feePercentage?: number,
+    executionGas?: number,
+    maxGasPrice?: string,
+    gasTokenAddress?: string,
+    salt?: string,
+    payAt?: number | null,
+    recurringDayOfMonth?: number | null,
+    recurringUntil?: number | null,
+    onScheduledPaymentCreate?: (scheduledPaymentId: string) => unknown
+  ) {
+    let hubAuth = await getSDK('HubAuth', this.web3);
+    let hubRootUrl = await hubAuth.getHubUrl();
+    let authToken = await hubAuth.authenticate();
+
+    let safeAddress: string;
+
+    if (isAddress(safeAddressOrScheduledPaymentId)) {
+      safeAddress = safeAddressOrScheduledPaymentId;
+    } else {
+      let scheduledPaymentId = safeAddressOrScheduledPaymentId;
+
+      return await waitUntilSchedulePaymentTransactionMined(hubRootUrl, scheduledPaymentId, authToken);
+    }
+
+    if (!moduleAddress) throw new Error('moduleAddress must be provided');
+    if (!tokenAddress) throw new Error('tokenAddress must be provided ');
+    if (!amount) throw new Error('amount must be provided');
+    if (!payeeAddress) throw new Error('payeeAddress must be provided');
+    if (!feeFixedUSD) throw new Error('feeFixedUSD must be provided');
+    if (!feePercentage) throw new Error('feePercentage must be provided');
+    if (!executionGas) throw new Error('executionGas must be provided');
+    if (!maxGasPrice) throw new Error('maxGasPrice must be provided');
+    if (!gasTokenAddress) throw new Error('gasTokenAddress must be provided ');
+    if (!salt) throw new Error('salt must be provided');
+    if (payAt == null && recurringDayOfMonth == null && recurringUntil == null)
+      throw new Error('When payAt is null, recurringDayOfMonth and recurringUntil must have a value');
+
+    let spHash: string;
+    if (recurringDayOfMonth && recurringUntil) {
+      spHash = await this.createSpHash(
+        moduleAddress,
+        tokenAddress,
+        amount,
+        payeeAddress,
+        { fixedUSD: feeFixedUSD, percentage: feePercentage },
+        executionGas,
+        maxGasPrice,
+        gasTokenAddress,
+        salt,
+        recurringDayOfMonth,
+        recurringUntil
+      );
+    } else {
+      spHash = await this.createSpHash(
+        moduleAddress,
+        tokenAddress,
+        amount,
+        payeeAddress,
+        { fixedUSD: feeFixedUSD, percentage: feePercentage },
+        executionGas,
+        maxGasPrice,
+        gasTokenAddress,
+        salt,
+        payAt!
+      );
+    }
+
+    let signature = await this.generateSchedulePaymentSignature(safeAddress, moduleAddress, gasTokenAddress, spHash);
```

Comment:
I think we can remove this line from this function and let the `schedulePaymentOnChain` function generate the signature.

I also think the `generateSchedulePaymentSignature` function can directly receive nonce, estimate, and payload as parameters, so the `gasEstimate` and `getNextNonceFromEstimate` won't be called twice in this flow.

Reply:
Great catch! Will fix it